real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/planners/WayPointFollowerPlanner.java

```diff
@@ -32,75 +49,219 @@ public void updatePositionEstimate(GPSPoseMessage m) {
     }
 
     //find the closest way point
-    //TODO turn into a binary search
-    private static int getClosestIndex(ArrayList<GpsMeasurement> wayPoints, GPSPoseMessage currentLocation) {
+    private int getClosestIndex(ArrayList<GpsMeasurement> wayPoints, GPSPoseMessage currentLocation) {
         double min = Double.MAX_VALUE; //note that the brakes will definitely deploy at this
 
-        int closestIndex = -1;
-        for (int i = 0; i < wayPoints.size(); i++) {
-            double d = GPSPoseMessage.getDistance(currentLocation, wayPoints.get(i).toGpsPoseMessage(0));
+        int closestIndex = lastClosestIndex;
+        for (int i = lastClosestIndex; i < (lastClosestIndex + WAYPOINT_LOOKAHEAD_MAX) && i < wayPoints.size(); i++) {
+            GPSPoseMessage gpsPoseMessage = wayPoints.get(i).toGpsPoseMessage(0);
+            double d = GPSPoseMessage.getDistance(currentLocation, gpsPoseMessage);
             if (d < min) {
                 min = d;
                 closestIndex = i;
             }
+            // eventually cut off search somehow
```

Comment:
not entirely necessary for this pull request, consider opening an issue/trello card for this